flink-python/src/main/java/org/apache/flink/python/metric/FlinkMetricContainer.java

```diff
@@ -19,17 +19,267 @@
+
 /**
  * Helper class for forwarding Python metrics to Java accumulators and metrics.
  */
 @Internal
-public class FlinkMetricContainer {
+public final class FlinkMetricContainer {
+
+	private static final String METRIC_KEY_SEPARATOR =
+		GlobalConfiguration.loadConfiguration().getString(MetricOptions.SCOPE_DELIMITER);
 
+	private final MetricsContainerStepMap metricsContainers;
 	private final MetricGroup baseMetricGroup;
+	private final Map<String, Counter> flinkCounterCache;
+	private final Map<String, Meter> flinkMeterCache;
+	private final Map<String, FlinkDistributionGauge> flinkDistributionGaugeCache;
+	private final Map<String, FlinkGauge> flinkGaugeCache;
 
 	public FlinkMetricContainer(MetricGroup metricGroup) {
-		this.baseMetricGroup = metricGroup;
+		this.baseMetricGroup = checkNotNull(metricGroup);
+		this.flinkCounterCache = new HashMap<>();
+		this.flinkMeterCache = new HashMap<>();
+		this.flinkDistributionGaugeCache = new HashMap<>();
+		this.flinkGaugeCache = new HashMap<>();
+		this.metricsContainers = new MetricsContainerStepMap();
+	}
+
+	private MetricsContainerImpl getMetricsContainer(String stepName) {
+		return metricsContainers.getContainer(stepName);
+	}
+
+	/**
+	 * Update this container with metrics from the passed {@link MonitoringInfo}s, and send updates
+	 * along to Flink's internal metrics framework.
+	 */
+	public void updateMetrics(String stepName, List<MonitoringInfo> monitoringInfos) {
+		getMetricsContainer(stepName).update(monitoringInfos);
+		updateMetrics(stepName);
+	}
+
+	/**
+	 * Update Flink's internal metrics ({@link this#flinkCounterCache}) with the latest metrics for
+	 * a given step.
+	 */
+	private void updateMetrics(String stepName) {
+		MetricResults metricResults = asAttemptedOnlyMetricResults(metricsContainers);
+		MetricQueryResults metricQueryResults =
+			metricResults.queryMetrics(MetricsFilter.builder().addStep(stepName).build());
+		updateCounterOrMeter(metricQueryResults.getCounters());
+		updateDistributions(metricQueryResults.getDistributions());
+		updateGauge(metricQueryResults.getGauges());
+	}
+
+	private boolean isUserMetric(MetricResult metricResult) {
+		MetricName metricName = metricResult.getKey().metricName();
+		return (metricName instanceof MonitoringInfoMetricName) &&
+			((MonitoringInfoMetricName) metricName).getUrn()
+				.contains(MonitoringInfoConstants.Urns.USER_COUNTER);
+	}
+
+	private void updateCounterOrMeter(Iterable<MetricResult<Long>> counters) {
+		for (MetricResult<Long> metricResult : counters) {
+			if (!isUserMetric(metricResult)) {
+				continue;
+			}
+			// get identifier
+			String flinkMetricIdentifier = getFlinkMetricIdentifierString(metricResult.getKey());
+
+			// get metric type
+			ArrayList<String> scopeComponents = getNameSpaceArray(metricResult.getKey());
+			if ((scopeComponents.size() % 2) != 0) {
+				Meter meter;
+				if (flinkMeterCache.containsKey(flinkMetricIdentifier)) {
```

Comment:
Change to the following then it will only lookup the map once:

counter = flinkCounterCache.get(flinkMetricIdentifier);
if (counter == null) {
   XXX
}